### PR TITLE
ROX-13888: Migrate permission sets referencing Policy and VulnerabilityReports resources

### PR DESCRIPTION
**Pull request by dhaus67**
## Description

This is the second and last PR in a series of PRs to remove the resources Policiy and VulnerabilityReports in favor of the WorkflowAdministration resource.

The PR itself covers the migration that touches all permission sets that reference the resource.

The migration follows the same logic as previously for resource removals: the lowest access level will be chosen of either the resource that should be removed or the replacing resource.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

- see CI passing.

Manual tests:
1. Add a test permission set that references both resources that will be replaced with this migration:
```bash
roxcurl /v1/permissionsets -X POST -d '{"id":"","name":"test-perm-set","description":"testing migration","resourceToAccess":{"Policy":"READ_ACCESS","VulnerabilityReports":"READ_WRITE_ACCESS"}}'
```

2. Update to the central image within this PR.

3. Verify within the logs that the migration has run:
```log
Migrator: 2023/06/07 14:25:37.853213 log.go:18: Info: Found DB at version 184, which is less than what we expect (185). Running migrations...
Migrator: 2023/06/07 14:25:37.864948 log.go:18: Info: Successfully updated DB from version 184 to 185
```

4. Verify within the API that the permission set is the expected one:
```bash
        {
            "id": "c8e24b11-a23c-44e4-8b74-83dc2131fb35",
            "name": "test-perm-set",
            "description": "testing migration",
            "resourceToAccess": {
                "WorkflowAdministration": "READ_ACCESS"
            },
            "traits": null
        }
```

### CODE REVIEW

File: pkg/migrations/internal/seq_num.go [hunk 1]
```diff
@@ -4,13 +4,13 @@ var (
...
 	// It is a shared constant between central and the migrator binary.
-	CurrentDBVersionSeqNum = 184
+	CurrentDBVersionSeqNum = 185
 
 	// MinimumSupportedDBVersionSeqNum is the minimum DB version number
 	// that is supported by this database.  This is used in case of rollbacks in
 	// the event that a major change introduced an incompatible schema update we
 	// can inform that a rollback below this is not supported by the database
-	MinimumSupportedDBVersionSeqNum = 184
+	MinimumSupportedDBVersionSeqNum = 185
```
Comment (dhaus67) [inline pkg/migrations/internal/seq_num.go:13]: @dashrews78 might be a dumb question, but: in cases like this where we only replace existing data sets but don't necessarily have schema changes, should this still be bumped or we just leave as-is?

Comment (dashrews78) [reply pkg/migrations/internal/seq_num.go:13]: @dhaus67 Not a dumb question.  In 4.1, yes we should bump that.  In 4.2 that should evolve such that we only need to set it for changes that break the compatibility of the database with the previous releases software.  But I had to put this piece in now so rollbacks from the future to 4.1 could be prevented if the future made schema changes that 4.1 could not support.  I have tickets to update the README and provide examples as part of my 4.2 work.

File: migrator/migrations/m_184_to_m_185_remove_policy_vulnerability_report_resources/migration_impl.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+				if !found {
+					accessForWorkflowAdministration = access
+				} else if access < accessForWorkflowAdministration {
+					// In case the permission set does have permission to the WorkflowAdministration resource _and_
+					// the current access for the resource to replace is lower than the one of the WorkflowAdministration,
+					// we will assign it the lower access.
+					// The reasoning behind is that we do not want to grant permission sets more permissions that they
+					// ideally require. The disadvantage is that this might impact customers that did not change their
+					// permission sets during the deprecation phase but want more permissions on the resource.
+					accessForWorkflowAdministration = access
```
[the remark this answers is not in the document]
Comment (dhaus67) [reply migrator/migrations/m_184_to_m_185_remove_policy_vulnerability_report_resources/migration_impl.go:52]: While I agree with the potential downsides in principle, this has already been communicated to the customers in the changelog and the same pattern was followed for all resource deprecations:

>The permission WorkflowAdministration will replace Policy, VulnerabilityReports in permission sets starting with the 4.1 release. You should preemptively start replacing the Policy and VulnerabilityReports resources within your permission sets in favor of WorkflowAdministration. During the migration of the permission sets within the 4.1, the WorfklowAdministration permission will have the lowest access permission granted for either Policy or VulnerabilityReports. As an example, a permission set with WRITE Policy and READ VulnerabilityReports access will have READ WorkflowAdministration access after the migration within the 4.1 release, leading to potentially unwanted side-effects and missing access if you did not update your permission sets beforehand.

The same argument could be made the other way around: when a customer only has `READ` on the report but `WRITE` on the policy, should he really have `WRITE` on `WorkflowAdministration`? Back then we chose the minimal access level approach, so this isn't something we can change now.

File: migrator/migrations/m_184_to_m_185_remove_policy_vulnerability_report_resources/migration_impl.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+				continue
+			}
+		}
+
+		if !hasReplacedResource {
+			return nil
+		}
+
+		permissionSetsToMigrate = append(permissionSetsToMigrate, obj)
+		if len(permissionSetsToMigrate) >= batchSize {
```
[the remark this answers is not in the document]
Comment (dhaus67) [reply migrator/migrations/m_184_to_m_185_remove_policy_vulnerability_report_resources/migration_impl.go:66]: The pattern is used throughout most (if not all migrations that do some upserting), and the tests seem fine so 🤷

File: migrator/migrations/m_184_to_m_185_remove_policy_vulnerability_report_resources/migration_impl.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+				} else if access < accessForWorkflowAdministration {
+					// In case the permission set does have permission to the WorkflowAdministration resource _and_
+					// the current access for the resource to replace is lower than the one of the WorkflowAdministration,
+					// we will assign it the lower access.
+					// The reasoning behind is that we do not want to grant permission sets more permissions that they
+					// ideally require. The disadvantage is that this might impact customers that did not change their
+					// permission sets during the deprecation phase but want more permissions on the resource.
+					accessForWorkflowAdministration = access
+				}
+				obj.ResourceToAccess[workflowAdministration] = accessForWorkflowAdministration
```
[the remark this answers is not in the document]
Comment (ivan-degtiarenko) [reply migrator/migrations/m_184_to_m_185_remove_policy_vulnerability_report_resources/migration_impl.go:54]: Nevermind, I see now

File: migrator/migrations/m_184_to_m_185_remove_policy_vulnerability_report_resources/migration_impl.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ivan-degtiarenko) [reply migrator/migrations/m_184_to_m_185_remove_policy_vulnerability_report_resources/migration_impl.go:52]: Ah, cool, didn't know that

**Review by ivan-degtiarenko — APPROVED** [review]

**Pull request merged**